File: 10-Recursion-and-Backtracking/java/05-RatInAMaze.java
class Solution {
    public List<String> findPath(int[][] grid) {
        List<String> result = new ArrayList<>();
        int n = grid.length;

        if (grid[0][0] == 0 || grid[n - 1][n - 1] == 0) {
            return result;
        }

        boolean[][] visited = new boolean[n][n];
        StringBuilder path = new StringBuilder();

        solve(0, 0, grid, n, visited, path, result);
        return result;
    }

    private static final int[] dx = {1, 0, 0, -1};
    private static final int[] dy = {0, -1, 1, 0};
    private static final char[] dirChar = {'D', 'L', 'R', 'U'};

    private void solve(int x, int y, int[][] grid, int n, boolean[][] visited,
                        StringBuilder path, List<String> result) {
        if (x == n - 1 && y == n - 1) {
            result.add(path.toString());
            return;
        }

        visited[x][y] = true;

        for (int dir = 0; dir < 4; dir++) {
            int newX = x + dx[dir];
            int newY = y + dy[dir];

            if (isSafe(newX, newY, grid, n, visited)) {
                path.append(dirChar[dir]);
                solve(newX, newY, grid, n, visited, path, result);
                path.deleteCharAt(path.length() - 1);
            }
        }

        visited[x][y] = false;
    }

    private boolean isSafe(int x, int y, int[][] grid, int n, boolean[][] visited) {
        return x >= 0 && x < n && y >= 0 && y < n
                && grid[x][y] == 1
                && !visited[x][y];
    }
}

File: 10-Recursion-and-Backtracking/cpp/05-rat-in-a-maze.cpp
class Solution {
    int dx[4] = {1, 0, 0, -1};
    int dy[4] = {0, -1, 1, 0};
    char dirChar[4] = {'D', 'L', 'R', 'U'};

public:
    vector<string> findPath(vector<vector<int>>& grid) {
        vector<string> result;
        int n = grid.size();

        if (grid[0][0] == 0 || grid[n - 1][n - 1] == 0) return result;

        vector<vector<bool>> visited(n, vector<bool>(n, false));
        string path = "";
        solve(0, 0, grid, n, visited, path, result);
        return result;
    }

private:
    void solve(int x, int y, const vector<vector<int>>& grid, int n,
               vector<vector<bool>>& visited, string& path, vector<string>& result) {
        if (x == n - 1 && y == n - 1) {
            result.push_back(path);
            return;
        }

        visited[x][y] = true;

        for (int dir = 0; dir < 4; dir++) {
            int newX = x + dx[dir];
            int newY = y + dy[dir];

            if (isSafe(newX, newY, grid, n, visited)) {
                path.push_back(dirChar[dir]);
                solve(newX, newY, grid, n, visited, path, result);
                path.pop_back();
            }
        }

        visited[x][y] = false;
    }

    bool isSafe(int x, int y, const vector<vector<int>>& grid, int n, const vector<vector<bool>>& visited) {
        return x >= 0 && x < n && y >= 0 && y < n && grid[x][y] == 1 && !visited[x][y];
    }
};
